src/charm_state.py

```diff
@@ -49,16 +19,25 @@ class State:
 
     Attrs:
         proxy_config: Proxy configuration.
+        _charm: The charm instance.
     """
 
-    proxy_config: Optional[ProxyConfig]
+    _charm: CharmBase
```

Comment:
I'm not a big fan of keeping the whole charm into the state, you should narrow down either to the relation or directly to the property (`is_metrics_logging_available`)